ts/components/conversation/media-gallery/groupMessagesByDate.ts

```diff
@@ -0,0 +1,164 @@
+  year: number;
+  month: number;
+};
+export type Section = StaticSection | YearMonthSection;
+
+export const groupMessagesByDate = (
+  timestamp: number,
+  messages: Array<Message>
+): Array<Section> => {
+  const referenceDateTime = moment.utc(timestamp);
+  const today = moment(referenceDateTime).startOf('day');
+  const yesterday = moment(referenceDateTime)
+    .subtract(1, 'day')
+    .startOf('day');
+  const thisWeek = moment(referenceDateTime).startOf('isoWeek');
+  const thisMonth = moment(referenceDateTime).startOf('month');
+
+  const sortedMessages = sortBy(messages, message => -message.received_at);
+  const messagesWithSection = sortedMessages.map(
+    withSection({
+      today,
+      yesterday,
+      thisWeek,
+      thisMonth,
+    })
+  );
+  const groupedMessages = groupBy(messagesWithSection, 'type');
+  const yearMonthMessages = Object.values(
+    groupBy(groupedMessages.yearMonth, 'order')
+  ).reverse();
+  return compact([
+    toSection(groupedMessages.today),
+    toSection(groupedMessages.yesterday),
+    toSection(groupedMessages.thisWeek),
+    toSection(groupedMessages.thisMonth),
+    ...yearMonthMessages.map(group => toSection(group)),
+  ]);
+};
+
+const toSection = (
+  messagesWithSection: Array<MessageWithSection> | undefined
+): Section | null => {
+  if (!messagesWithSection || messagesWithSection.length === 0) {
+    return null;
+  }
+
+  const firstMessageWithSection: MessageWithSection = messagesWithSection[0];
+  if (!firstMessageWithSection) {
+    return null;
+  }
+
+  const messages = messagesWithSection.map(
+    messageWithSection => messageWithSection.message
+  );
+  switch (firstMessageWithSection.type) {
+    case 'today':
+    case 'yesterday':
+    case 'thisWeek':
+    case 'thisMonth':
+      return {
+        type: firstMessageWithSection.type,
+        messages,
+      };
+    case 'yearMonth':
+      return {
+        type: firstMessageWithSection.type,
+        year: firstMessageWithSection.year,
+        month: firstMessageWithSection.month,
+        messages,
+      };
+    default:
+      // NOTE: Investigate why we get the following error:
+      // error TS2345: Argument of type 'any' is not assignable to parameter
+      // of type 'never'.
+      // return missingCaseError(firstMessageWithSection.type);
+      return null;
+  }
+};
+
+interface GenericMessageWithSection<T> {
+  order: number;
+  type: T;
+  message: Message;
+}
+type MessageWithStaticSection = GenericMessageWithSection<StaticSectionType>;
+type MessageWithYearMonthSection = GenericMessageWithSection<
+  YearMonthSectionType
+> & {
+  year: number;
+  month: number;
+};
+type MessageWithSection =
+  | MessageWithStaticSection
+  | MessageWithYearMonthSection;
+
+const withSection = ({
+  today,
+  yesterday,
+  thisWeek,
+  thisMonth,
+}: {
+  today: moment.Moment;
+  yesterday: moment.Moment;
+  thisWeek: moment.Moment;
+  thisMonth: moment.Moment;
+}) => (message: Message): MessageWithSection => {
```

Comment:
Ah, here's the original algorithm. :0) Why did all of the timings need to be passed in, if they are calculated based on a provided `now` value?